### DISCUSSION TITLE
Confine bodies inside a convex polygon or circle.

### DISCUSSION THREAD

**Discussion by mbrotz** [dyn4j/dyn4j#282] [General]
Hi

I am using dyn4j in a processing project and would like to confine several objects inside a convex polygon or a circle. Currently i am using links to achieve that. It works to some extent, but some objects sometimes tunnel through the "walls". (I disabled CCD and would like to keep it that way.)

Can you give me some advice on how to avoid the tunneling problem? The objects are never supposed to leave the confinement, so even when they tunnel through the wall and don't intersect the line segments, the collision should still be detected. Essentially, as if the walls were infinitely thick, expanding to the outside forever. 

Thanks!
Manuel

**Comment by wnbittle** [comment]
The tunneling problem is a difficult one and CCD can only do so much too.  If you don't like the way CCD works in dyn4j (time-stealing), then understanding why tunneling happens is important to know what your options are.

Tunneling occurs due to one reason - the collision detection system (really the whole simulation) works on fixed timesteps (also called discrete collision detection).  What this means is that we know what the state of the system is at `T0` and at `T1`, but nothing about what happened in between.  If the time between `T0` and `T1` is large, lots of things can happen without us knowing - things like objects passing through each other.

So the natural solution would be to lower your timestep by `world.getSettings().setStepFrequency(...)`.  The default frequency is `1.0/60.0` which means 60 times per second.  A larger denominator means the whole simulation runs more often.  This also means that the time between `T0` and `T1` is smaller.  Because the time between `T0` and `T1` is smaller, we miss less things.  Some simulations can support this, but it should be obvious that running the simulation more frequently will have a performance impact.

Changing the step frequency _may_ work, but you can _still_ experience tunneling.  Why?  Think about it, no matter how small the time is between `T0` and `T1`, we can always construct a situation where a `Body` moves fast enough that a collision is still missed.  You could avoid this by setting the `world.getSettings().setMaximumTranslation(...)` and `world.getSettings().setMaximumRotation(...)` which puts an upper limit on the amount a `Body` can move/rotate in a given step.

Changing both the step frequency _and_ the maximum translation/rotation _may_ work, but you can _still_ experience tunneling.  Why? Think about it, no matter how small the time is between `T0` and `T1` AND no matter how much I slow down a `Body`, I could still construct a situation where the `Body`s are so small that even when they go slow, the collision is missed - consider a `Circle` with a radius close to zero and an infinitely thin `Segment`.  This is the trickiest one, you have to make your objects _big enough_.

With this information, here are a few things you can try:

1. Increase the simulation rate using `world.getSettings().setStepFrequency(...)` with values smaller than 1/60
2. Clamp the maximum speed of your `Body`s
3. Make sure your `Body`s are of sufficient size for collisions to be detected
4. Create a new `Shape` which resembles an infinitely deep object (similar to a plane in 3D) - this is the same as the previous point, just a different way of achieving the same thing

Here's a (terrible) example of using larger `Body`s to avoid tunneling (instead of `Segment`/`Link`s):
<img src="https://github.com/dyn4j/dyn4j/assets/11467997/287c19f4-08de-427e-be4e-4cb9aa42a825" width="300"/>

This is done in a number of the [samples project](https://github.com/dyn4j/dyn4j-samples) simulations.

**Reply by mbrotz** [reply]
Hi! 

Thanks for your answer and sorry for the long delay.

I agree with everything you said. One of the reasons i don't use larger bodys was because of the smooth sliding problem. But i suppose this problem goes away if i make the large bodies overlap enough...

It would be cool to support a special type of links where the line segments are not treated as infinitely thin but instead infinitely deep... Do you think that's possible or are there some fundamental problems with that? I could try to implement that myself.

**Reply by wnbittle** [reply]
> I agree with everything you said. One of the reasons i don't use larger bodys was because of the smooth sliding problem. But i suppose this problem goes away if i make the large bodies overlap enough...

For the most part, that's true.  There's still some art in defining your shapes so that they are good for smooth sliding, handle fast collisions well, and relatively simple.

> It would be cool to support a special type of links where the line segments are not treated as infinitely thin but instead infinitely deep... Do you think that's possible or are there some fundamental problems with that? I could try to implement that myself.

There are two challenges that come to mind with this approach:

1: If segments extend infinitely on one side, they have the possibility to overlap.  When they overlap, they'll cause some fun behavior because the engine may think an object penetrated into the segment super deep and send it flying to get it out.  For example, consider this drawing of a chain of `Link` shapes.  

![image](https://github.com/dyn4j/dyn4j/assets/11467997/17986ac5-e157-4113-bad8-c860437c0e8c)

Each link extends infinitely on one side.  Things look good until you get to the C-D segment - the E-F segment's infinite extension overlaps C-D partially and D-E completely.  If an object attempted to enter that green area, it would be stopped right after the C point.  Not only that, the engine would think that the object tunneled all the way down to that point (rather than entering from the side).

2: The other issue, which is a little harder to catch is that the junction between these segments is still a problem even if you extend them infinitely.  For example, in the image above, the point B.  If an object is sufficiently small and/or fast, it can tunnel through that point because there's still an open area where we wouldn't detect collision.

One option might be to extend the line segments to lines and make those infinitely deep, but you run into more of problem 1.

> I am using dyn4j in a processing project and would like to confine several objects inside a convex polygon or a circle

Now, that said - if you wanted to solve the very specific problem of a containing an object in a **Convex** polygon, you _can_ extend the segments into lines _and_ extend one side infinitely without running into issues.  For example:

![image](https://github.com/dyn4j/dyn4j/assets/11467997/4a5e299b-bf37-4f52-a69d-8aad1ee6bfe0)

Thinking about this critically, it's an adaptation of the the SAT algorithm.  You would project the shape onto each line and test for "overlap" (it would be testing a fixed interval with an infinite interval).

**Comment by mbrotz** [comment]
Thanks so much for your explanations!

I thought about the link based version and came to the conclusion that it would be difficult to do it. The convex polygon version that you pointed out is very interesting though and exactly what i need! I will take a look at it. Maybe i can implement that.

It would be a great feature to add to the engine, IMO... ;)
